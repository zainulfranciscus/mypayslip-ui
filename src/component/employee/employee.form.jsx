import React, {Component} from 'react';

export const ID_OF_FIRST_NAME_TEXT_FIELD = "firstName";
export const ID_OF_LAST_NAME_TEXT_FIELD = "lastName";
export const ID_OF_ANNUAL_SALARY_TEXT_FIELD = "salary";
export const ID_OF_SUPER_RATE_TEXT_FIELD = "superRate";
export const ID_OF_GENERATE_PAYSLIP_BUTTON = "generatePayslipButton";
export const CLASS_NAME_FOR_INVALID_FIELD_ENTRY = "is-invalid";
export const EMPLOYEE_FORM_CONTAINER = "employeeForm";

export default class EmployeeForm extends Component {


    constructor(props) {
        super(props);
        let formValues = {};
        formValues[ID_OF_FIRST_NAME_TEXT_FIELD] = '';
        formValues[ID_OF_LAST_NAME_TEXT_FIELD] = '';
        formValues[ID_OF_ANNUAL_SALARY_TEXT_FIELD] = '';
        formValues[ID_OF_SUPER_RATE_TEXT_FIELD] = '';

        this.state = {
            formValues: formValues,
            isFirstNameValid: false,
            isLastNameValid: false,
            isSalaryValid: false,
            isSuperValid: false
        }


    }

    handleChange(event) {

        const name = event.target.name;
        const value = event.target.value;

        let updatedFormValues = this.state.formValues;
        updatedFormValues[name] = value;
        this.setState({formValues: updatedFormValues});

    }

    validateFirstName(event) {
        this.handleChange(event);
        let isValidFirstName = this.props.validation.withFirstName(event.target.value).validateFirstName();
        this.setState({isFirstNameValid: isValidFirstName});
    }

    validateLastName(event) {
        this.handleChange(event);
        let isValidLastName = this.props.validation.withLastName(event.target.value).validateLastName();
        this.setState({isLastNameValid: isValidLastName});
    }

    validateSalary(event) {
        this.handleChange(event);
        let isValidSalary = this.props.validation.withSalary(event.target.value).validateSalary();
        this.setState({isSalaryValid: isValidSalary});
    }

    validateSuper(event) {
        this.handleChange(event);
        let isValidSuper = this.props.validation.withSuper(event.target.value).validateSuper();
        this.setState({isSuperValid: isValidSuper});
    }

    generatePayslip() {
        let validation = this.props.validation.validate();
        if (validation.isEmployeeDataValid()) {
            this.props.generatePayslip(this.state.formValues);
        }
    }

    render() {

        return (
            <div id={EMPLOYEE_FORM_CONTAINER}>

                <h1>Employee Info</h1>
                <div className="form-group row">
                    
                        <div className="col-12 col-md-4">
                            <input type="text" id={ID_OF_FIRST_NAME_TEXT_FIELD}
                                   name={ID_OF_FIRST_NAME_TEXT_FIELD}
                                   onChange={this.validateFirstName.bind(this)}
                                   value={this.state.formValues[ID_OF_FIRST_NAME_TEXT_FIELD]}
                                   placeholder="first name"
                                   className={
                                       this.state.isFirstNameValid ? "form-control" : CLASS_NAME_FOR_INVALID_FIELD_ENTRY + " form-control"
                                   }/>


                            <p id="errorMessageforFirstName" className={
                                this.state.isFirstNameValid ? "d-none" : "error-msg"
                            }>Please enter a first name. First name cannot be a number </p>

                        </div>
                        <div className="col-12 col-md-4">
                            <input type="text" id={ID_OF_LAST_NAME_TEXT_FIELD}
                                   name={ID_OF_LAST_NAME_TEXT_FIELD}
                                   onChange={this.validateLastName.bind(this)}
                                   value={this.state.formValues[ID_OF_LAST_NAME_TEXT_FIELD]}
                                   placeholder="last name"
                                   className={
                                       this.state.isLastNameValid ? "form-control" : CLASS_NAME_FOR_INVALID_FIELD_ENTRY + " form-control"
                                   }/>
                            <p id="errorMessageforLastName" className={
                                this.state.isLastNameValid ? "d-none" : "error-msg"
                            }>Please enter a last name. Last name cannot be a number </p>

                        </div>
                    

                </div>
                
                    <div className="form-group row">
                        <div className="col-12 col-md-4">
                           <div className="input-group">
                                <div className="input-group-prepend">
                                    <div className="input-group-text">$</div>
                                </div>
                                <input type="text" id={ID_OF_ANNUAL_SALARY_TEXT_FIELD}
                                       name={ID_OF_ANNUAL_SALARY_TEXT_FIELD}
                                       onChange={this.validateSalary.bind(this)}
                                       value={this.state.formValues[ID_OF_ANNUAL_SALARY_TEXT_FIELD]}
                                       placeholder="Annual Salary"
                                       className={
                                           this.state.isSalaryValid ? "form-control" : CLASS_NAME_FOR_INVALID_FIELD_ENTRY + " form-control"
                                       }/>
                                <div className="input-group-prepend">
                                    <div className="input-group-text">.00</div>
                                </div>
                            </div>
                             <p id="errorMessageForSalary" className={
                                this.state.isSalaryValid ? "d-none" : "error-msg"
                            }>Please enter a salary, which must be a round number </p>
                           

                        </div>
                        <div className="col-12 col-md-4">
                            <input type="text" id={ID_OF_SUPER_RATE_TEXT_FIELD}
                                   name={ID_OF_SUPER_RATE_TEXT_FIELD}
                                   onChange={this.validateSuper.bind(this)}
                                   value={this.state.formValues[ID_OF_SUPER_RATE_TEXT_FIELD]}
                                   placeholder="Super Rate"
                                   className={
                                       this.state.isSuperValid ? "form-control" : CLASS_NAME_FOR_INVALID_FIELD_ENTRY + " form-control"
                                   }/>
                             <p id="errorMessageForSuper" className={
                                this.state.isSuperValid ? "d-none" : "error-msg"
                            }>Please enter a number for super. </p>
                        </div>
                    </div>

                    <div className="row">
                        <div className="input-group col-12 col-md-4">

                           
                        </div>
                        <div className="input-group col-12 col-md-4">
                           
                        </div>
                    </div>
                
                <div className="row justify-content-end">
                    <div className="input-group col-6 button-padding">
                        <button id={ID_OF_GENERATE_PAYSLIP_BUTTON}
                                name={ID_OF_GENERATE_PAYSLIP_BUTTON}
                                onChange={this.handleChange.bind(this)}
                                onClick={this.generatePayslip.bind(this)}
                                className="btn btn-primary "
                                value={this.state.formValues[ID_OF_GENERATE_PAYSLIP_BUTTON]}>Generate Payslip
                        </button>
                    </div>
                </div>


            </div>
        );
    }
}