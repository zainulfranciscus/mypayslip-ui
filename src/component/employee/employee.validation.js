export default class EmployeeValidation {

    constructor() {
        this.firstName = "";
        this.lastName = "";
        this.superRate = "";
        this.salary = "";
    }

    isNumeric(number) {
        return !isNaN(number);
    }

    isRoundNumber(number) {

        return this.isNumeric(number) && number % 1 === 0;
    }

    isEmpty(str) {
        str = str.trim();
        return (!str || 0 === str.length);
    }

    withSalary(salary) {
        this.salary = salary;
        return this;
    }

    withFirstName(firstName) {
        this.firstName = firstName;
        return this;
    }

    withLastName(lastName) {
        this.lastName = lastName;
        return this;
    }

    withSuper(superRate) {
        this.superRate = superRate;
        return this;
    }

    validateSalary() {
        return !this.isEmpty(this.salary) && this.isRoundNumber(this.salary) && parseFloat(this.salary) >= 0;
    }

    validateFirstName() {
        return (!this.isEmpty(this.firstName) && !this.isNumeric(this.firstName));
    }

    validateLastName() {
        return (!this.isEmpty(this.lastName) && !this.isNumeric(this.lastName));
    }

    validateSuper() {
        return !this.isEmpty(this.superRate) && this.isNumeric(this.superRate) && parseFloat(this.superRate) >= 0;
    }

    validate() {

        let firstNameValidationResult = this.validateFirstName();
        let lastNameValidationResult = this.validateLastName();
        let salaryValidationResult = this.validateSalary();
        let superValidationResult = this.validateSuper();

        return {
            firstNameValidation: firstNameValidationResult,
            lastNameValidation: lastNameValidationResult,
            salaryValidation: salaryValidationResult,
            superValidation: superValidationResult,
            isEmployeeDataValid: () => {
                let isValid = (firstNameValidationResult &&
                    lastNameValidationResult &&
                    salaryValidationResult &&
                    superValidationResult);

                return isValid;
            }
        };

    }
}