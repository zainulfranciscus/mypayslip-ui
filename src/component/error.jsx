import React, {Component} from 'react';
import taxDecorator from "../data/tax.decorator";
import payslipDecorator from "../data/payslip.decorator";

export const ERROR_MESSAGE_ID = "errorMessage";
export const SERVER_IS_UNAVAILABLE = "Server is unavailable to generate payslip";
export const SERVER_RETURNED_NO_TAX_DATA = "Server is unable to find a tax bracket for this employee. Payslip will not be generated";
export const SERVER_RETURNED_INVALID_TAX_DATA = "Tax bracket returned by the server is missing the minimum information to calculate a payslip.";
export default class Error extends Component {

    exist(aProp) {
        return (typeof aProp !== 'undefined');
    }

    isPostPayslipThrewAnError() {
        const wasErrorThrown = this.props.postPayslipError;
        return this.exist(wasErrorThrown) && this.props.postPayslipError;

    }

    isFetchTaxThrewAnError() {
        const wasErrorThrown = this.props.fetchTaxError;
        return this.exist(wasErrorThrown) && this.props.fetchTaxError;

    }

    errorDiv(message) {
        return (
            <div id={ERROR_MESSAGE_ID} className="row error-container">
                <div className="col-12">
                    {message}
                </div>
            </div>
        );
    }

    render() {


        const taxResponseJSON = taxDecorator(this.props.taxResponseJSON);
        const payslipResponseJSON = payslipDecorator(this.props.payslipResponseJSON);

        if (this.isFetchTaxThrewAnError() || this.isPostPayslipThrewAnError()) {
            return this.errorDiv(SERVER_IS_UNAVAILABLE);
        }


        if (taxResponseJSON.jsonExist() && !taxResponseJSON.ok()) {
            return this.errorDiv(taxResponseJSON.message());
        }

        if (taxResponseJSON.jsonExist() && !taxResponseJSON.hasValidTaxData()) {
            return this.errorDiv(SERVER_RETURNED_INVALID_TAX_DATA);
        }

        if (payslipResponseJSON.jsonExist() && !payslipResponseJSON.ok() && payslipResponseJSON.hasMessage()) {
            return this.errorDiv(payslipResponseJSON.message());
        }

        return null;
    }
}