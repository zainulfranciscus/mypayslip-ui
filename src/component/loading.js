import React, {Component} from 'react';

export const GETTING_TAX = "Please wait while I fetch tax information for this employee";
export const SAVING_PAYSLIP = "Please wait while I save this employee's payslip";
export const ID_OF_LOADER = "loader";

export default class Loading extends Component {

    getLoadingDiv(message){
        return (
            <div className="row loading-component" id={ID_OF_LOADER}>
                <div className="col-6">
                    {message}
                </div>
            </div>
        )
    }
    render() {
        const isLoadingTax = this.props.loadingTax;
        const isSavingPayslip = this.props.isSavingPayslip;


        if(this.props.hasError){
            return null;
        }

        if (isLoadingTax) {
            return this.getLoadingDiv(GETTING_TAX);
        }

        if (isSavingPayslip) {
            return this.getLoadingDiv(SAVING_PAYSLIP);
        }

        return null;
    }
}
