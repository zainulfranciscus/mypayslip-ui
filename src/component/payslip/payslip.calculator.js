import {
    ID_OF_ANNUAL_SALARY_TEXT_FIELD,
    ID_OF_SUPER_RATE_TEXT_FIELD,
    ID_OF_FIRST_NAME_TEXT_FIELD,
    ID_OF_LAST_NAME_TEXT_FIELD
} from '../employee/employee.form';
import taxDecorator from '../../data/tax.decorator';

export const NUMBER_OF_MONTHS_IN_A_YEAR = 12;

export default class PayslipCalculator {

    fullName(){

        let firstName = this.firstName();
        let lastName = this.lastName();

        if(typeof firstName !== 'undefined' && firstName !== null){
            firstName = firstName.trim();
        }

        if(typeof lastName !== 'undefined' && lastName !== null){
            lastName = lastName.trim();
        }

        return firstName + " " + lastName;
    }

    firstName(){
        return this.employeeFormValues[ID_OF_FIRST_NAME_TEXT_FIELD];
    }

    lastName(){
        return this.employeeFormValues[ID_OF_LAST_NAME_TEXT_FIELD];
    }

    annualSalary() {
        return parseFloat(this.employeeFormValues[ID_OF_ANNUAL_SALARY_TEXT_FIELD]);
    }

    superRate() {
        return parseFloat(this.employeeFormValues[ID_OF_SUPER_RATE_TEXT_FIELD]) / 100;
    }

    setEmployee(employee) {
        this.employeeFormValues = employee;
    }

    setTax(taxJson) {
        this.tax = taxDecorator(taxJson);
    }

    roundUpAboveZeroPoint5(aNumber) {
        return Math.round(aNumber, -1);
    }

    getGrossIncome() {
        const annualSalary = this.annualSalary();
        return this.roundUpAboveZeroPoint5(divide(annualSalary).by(NUMBER_OF_MONTHS_IN_A_YEAR));
    }

    getIncomeTax() {
        const yearlyTax = this.getYearlyTax();
        return this.roundUpAboveZeroPoint5(divide(yearlyTax).by(NUMBER_OF_MONTHS_IN_A_YEAR));
    }

    getNetIncome() {
        return this.getGrossIncome() - this.getIncomeTax();
    }

    getSuper() {
        const amountOfSuper = this.superRate();
        const grossIncome = this.getGrossIncome();

        return this.roundUpAboveZeroPoint5(grossIncome * amountOfSuper);
    }

    getPay() {
        return this.getNetIncome() - this.getSuper();
    }

    getTaxableIncome() {
        const taxPerDollarOver = this.tax.getTaxPerDollarOver();
        return this.annualSalary() - taxPerDollarOver;
    }

    getTaxForEachTaxableDollar() {
        const taxPerDollar = this.tax.getTaxPerDollarInCents();
        return this.getTaxableIncome() * taxPerDollar;
    }

    getYearlyTax() {
        const baseTax = this.tax.getBaseTax();
        return (baseTax + this.getTaxForEachTaxableDollar());
    }

}


export function divide(aNumber) {
    return {
        by: (divisor) => {
            return aNumber / divisor;
        }
    }
}