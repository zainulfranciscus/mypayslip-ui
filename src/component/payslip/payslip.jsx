import React, {Component} from 'react';
import moment from "moment";
import payslipDecorator from "../../data/payslip.decorator";

const ID_FOR_PAYDATE_DIV = "Paydate_ID";
const ID_FOR_PAYDATE_LABEL = "Paydate_Label";
const PAYDATE_LABEL = "Pay Date";

const ID_FOR_PAY_FREQUENCY = "PayFrequency_ID";
const ID_FOR_PAY_FREQUENCY_LABEL = "PayFrequency_Label";
const FREQUENCY_LABEL = "Pay Frequency";
const PAY_FREQUENCY_VALUE = "Monthly";

const ID_FOR_ANNUAL_INCOME = "AnnualIncome_ID";
const ID_FOR_ANNUAL_INCOME_LABEL = "AnnualIncome_Label";
const ANNUAL_INCOME_LABEL = "Annual Income";

const ID_FOR_GROSS_INCOME = "GrossIncome_ID";
const ID_FOR_GROSS_INCOME_LABEL = "GrossIncome_Label";
const GROSS_INCOME_LABEL = "Gross Income";


const ID_FOR_INCOME_TAX = "IncomeTax_ID";
const ID_FOR_INCOME_TAX_LABEL = "IncomeTax_Label";
const INCOME_TAX_LABEL = "Income Tax";

const ID_FOR_NET_INCOME = "NetIncome_ID";
const ID_FOR_NET_INCOME_LABEL = "NetIncome_Label";
const NET_INCOME_LABEL = "Net Income";

const ID_FOR_SUPER = "Super_ID";
const ID_FOR_SUPER_LABEL = "Super_Label";
const SUPER_LABEL = "Super";

const ID_FOR_PAY = "Pay_ID";
const ID_FOR_PAY_LABEL = "Pay_Label";
const PAY_LABEL = "Pay";


export const ID_FOR_PAYSLIP_CONTAINER = "Payslip_Container";

export const DATE_FORMAT = 'DD MMM YYYY';


export default class Payslip extends Component {

    postPayslipToServer() {

        const calculator = this.props.calculator;

        const payslipRequestParameters = {
            annualSalary: calculator.annualSalary(),
            grossIncome:calculator.getGrossIncome(),
            netIncome: calculator.getNetIncome(),
            superRate: calculator.getSuper(),
            pay:calculator.getPay(),
            firstName:calculator.firstName(),
            lastName:calculator.lastName(),
            incomeTax: calculator.getIncomeTax()
        };

        this.props.savePayslip(payslipRequestParameters);
    }

    payslipHasBeenSaved() {
        const responseFromServer = this.props.payslipResponseFromServer;
        let decoratedPayslipResponse = payslipDecorator(responseFromServer);

        if (!decoratedPayslipResponse.jsonExist() || !decoratedPayslipResponse.ok()) {
            return null;
        }

        if(this.props.calculator.fullName() !== decoratedPayslipResponse.fullName()) {
            return null;
        }

        const calculator = this.props.calculator;
        const msg = "Payslip for " + calculator.fullName() + " has been saved";

        return (
            <div className="row payslip-saved">
                <div className="col-12">
                   {msg}
                </div>
            </div>
        );

    }

    render() {
        
        const payslipHasBeenSaved = this.payslipHasBeenSaved();

        if(typeof payslipHasBeenSaved !== 'undefined' && payslipHasBeenSaved !== null){
            return payslipHasBeenSaved;
        }

        return (
            <div id={ID_FOR_PAYSLIP_CONTAINER}>

                <h1>Payslip</h1>

                <p>{this.props.calculator.fullName()}</p>
                
                <div className="row grey-row">
                    <div className="col-6">
                        <div id={ID_FOR_PAYDATE_LABEL}><strong>{PAYDATE_LABEL}</strong></div>
                    </div>
                    <div className="col-6">
                        <div id={ID_FOR_PAYDATE_DIV}>{moment().format(DATE_FORMAT)}</div>
                    </div>
                </div>

                <div className="row">
                    <div className="col-6">
                        <div id={ID_FOR_PAY_FREQUENCY_LABEL}><strong>{FREQUENCY_LABEL}</strong></div>
                    </div>
                    <div className="col-6">
                        <div id={ID_FOR_PAY_FREQUENCY}>{PAY_FREQUENCY_VALUE}</div>
                    </div>
                </div>

                <div className="row grey-row">
                    <div className="col-6">
                        <div id={ID_FOR_ANNUAL_INCOME_LABEL}><strong>{ANNUAL_INCOME_LABEL}</strong></div>
                    </div>
                    <div className="col-6">
                        <div id={ID_FOR_ANNUAL_INCOME}>{this.props.calculator.annualSalary()}</div>
                    </div>
                </div>

                <div className="row">
                    <div className="col-6">
                        <div id={ID_FOR_GROSS_INCOME_LABEL}><strong>{GROSS_INCOME_LABEL}</strong></div>
                    </div>
                    <div className="col-6">
                        <div id={ID_FOR_GROSS_INCOME}>{this.props.calculator.getGrossIncome()}</div>
                    </div>
                </div>

                <div className="row grey-row">
                    <div className="col-6">
                        <div id={ID_FOR_INCOME_TAX_LABEL}><strong>{INCOME_TAX_LABEL}</strong></div>
                    </div>
                    <div className="col-6">
                        <div id={ID_FOR_INCOME_TAX}>{this.props.calculator.getIncomeTax()}</div>
                    </div>
                </div>

                <div className="row">
                    <div className="col-6">
                        <div id={ID_FOR_NET_INCOME_LABEL}><strong>{NET_INCOME_LABEL}</strong></div>
                    </div>
                    <div className="col-6">
                        <div id={ID_FOR_NET_INCOME}>{this.props.calculator.getNetIncome()}</div>
                    </div>
                </div>

                <div className="row grey-row">
                    <div className="col-6">
                        <div id={ID_FOR_SUPER_LABEL}><strong>{SUPER_LABEL}</strong></div>
                    </div>
                    <div className="col-6">
                        <div id={ID_FOR_SUPER}>{this.props.calculator.getSuper()}</div>
                    </div>
                </div>

                <div className="row">
                    <div className="col-6">
                        <div id={ID_FOR_PAY_LABEL}><strong>{PAY_LABEL}</strong></div>
                    </div>
                    <div className="col-6">
                        <div id={ID_FOR_PAY}>{this.props.calculator.getPay()}</div>
                    </div>
                </div>


                <div className="row justify-content-end">
                    <div className="col-6">
                        <button id="payButton" onClick={this.postPayslipToServer.bind(this)} className="btn btn-primary pay-btn">Pay
                        </button>
                    </div>
                </div>
                              
            </div>
        );
    }
}