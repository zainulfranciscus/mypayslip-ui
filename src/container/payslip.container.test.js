import React from "react";
import PayslipContainer from "./payslip.container";
import {ERROR_MESSAGE_ID} from "../component/error";
import {EMPLOYEE_FORM_CONTAINER} from "../component/employee/employee.form";
import {ID_FOR_PAYSLIP_CONTAINER} from "../component/payslip/payslip";
import {shallow, mount} from 'enzyme';
import configureStore from 'redux-mock-store';
import PayslipCalculator from "../component/payslip/payslip.calculator";
import {setupMockCalculatorMethods} from "../component/payslip/payslip.test";
import {ID_OF_LOADER} from "../component/loading";

describe("unit tests for PayslipContainer", () => {

    const mockStore = configureStore();
    let store;
    const mockCalculator = new PayslipCalculator();
    const mockItems = {
        tax: {
            status: 200,
            responseObject: {
                "baseTax": "540.00",
                "taxPerDollarInCents": "32.50",
                "taxPerDollarOver": "3752.00",
                "maxTaxableIncome": "18200.00",
                "minTaxableIncome": "10.00"
            }
        },
        employeeFormValues: {}

    };

    const savingPayslipResponse = {
        status: 400,
        message: "payslip for this person already exist"
    };

    const findErrorDiv = (wrapper) => {
        return wrapper.find("#" + ERROR_MESSAGE_ID);
    };

    beforeEach(() => {
        setupMockCalculatorMethods(mockCalculator);
    });
    it("when isLoading props is true then loader should be rendered ", () => {


        store = mockStore({loadingTax: true, items: mockItems});
        const wrapper = mount(<PayslipContainer store={store} calculator={mockCalculator} />);
        expect(wrapper.find("#" + ID_OF_LOADER).length).toBe(1);
    });

    it("when an error occured when retrieving tax for an employee, then error message shold be rendered ", () => {

        store = mockStore({
            errorWhenGettingTax: true
        });
        const wrapper = mount(<PayslipContainer store={store}  />);
        expect(findErrorDiv(wrapper).length).toBe(1);

    });

    it("when an error occured when saving a payslip, then error message should be rendered ", () => {

        store = mockStore({
            postingPayslipHasErrored: {
                hasErrored: true,
                employeeFormValues: {}
            }
        });
        const wrapper = mount(<PayslipContainer store={store}/>);
        expect(findErrorDiv(wrapper).length).toBe(1);

    });

    it("when rest API returns a status of 400 after saving a payslip, then  error message should be rendered", () =>{

        store = mockStore({
            payslip: {
                status: 400,
                message: 'Unable to save payslip',
                responseObject: {}
            }
        });

        const wrapper = mount(<PayslipContainer store={store}/>);
        expect(findErrorDiv(wrapper).length).toBe(1);

    });

    it("when rest API returns a status of 400 after fetching a tax, then  error message should be rendered", () =>{

        store = mockStore({
            items:{
                tax: {
                    status: 400,
                    message: 'Salary for an employee must be more than 0',
                    responseObject: {}
                }
            }
            
        });

        const wrapper = mount(<PayslipContainer store={store}/>);
        expect(findErrorDiv(wrapper).length).toBe(1);

    });

    it("when hasErrored is false and isLoading is false then employee form should be rendered", () => {


        store = mockStore({
            errorWhenGettingTax: false,
            isLoading: false,
            items: mockItems,
            payslip: {savingPayslipResponse}
        });
        const wrapper = mount(<PayslipContainer store={store} calculator={mockCalculator}/>);
        expect(wrapper.find("#" + EMPLOYEE_FORM_CONTAINER).length).toBe(1);

    });

    it("when employeeFormValues is set and tax is set then payslip should be rendered", () => {
        store = mockStore({
            items: mockItems,
            payslip: {savingPayslipResponse}
        });
        const wrapper = mount(<PayslipContainer store={store} calculator={mockCalculator}/>);
        expect(wrapper.find("#" + ID_FOR_PAYSLIP_CONTAINER).length).toBe(1);
    });

    it("when employeeFormValues and tax is undefined then payslip should NOT be rendered", () => {
        store = mockStore({
            items: {
                tax: undefined,
                employeeFormValues: undefined
            }
        });
        const wrapper = mount(<PayslipContainer store={store} calculator={mockCalculator}/>);
        expect(wrapper.find("#" + ID_FOR_PAYSLIP_CONTAINER).length).toBe(0);
    })


});