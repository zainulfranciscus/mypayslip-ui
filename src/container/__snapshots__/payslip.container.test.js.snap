// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Payslip component unit test payslip component should match the most up to date snapshot 1`] = `
<div
  id="Payslip_Container"
>
  <h1>
    Payslip
  </h1>
  <p>
    John Doe
  </p>
  <div
    className="row grey-row"
  >
    <div
      className="col-6"
    >
      <div
        id="Paydate_Label"
      >
        <strong>
          Pay Date
        </strong>
      </div>
    </div>
    <div
      className="col-6"
    >
      <div
        id="Paydate_ID"
      >
        26 April 2018
      </div>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col-6"
    >
      <div
        id="PayFrequency_Label"
      >
        <strong>
          Pay Frequency
        </strong>
      </div>
    </div>
    <div
      className="col-6"
    >
      <div
        id="PayFrequency_ID"
      >
        Monthly
      </div>
    </div>
  </div>
  <div
    className="row grey-row"
  >
    <div
      className="col-6"
    >
      <div
        id="AnnualIncome_Label"
      >
        <strong>
          Annual Income
        </strong>
      </div>
    </div>
    <div
      className="col-6"
    >
      <div
        id="AnnualIncome_ID"
      >
        $2,500.00
      </div>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col-6"
    >
      <div
        id="GrossIncome_Label"
      >
        <strong>
          Gross Income
        </strong>
      </div>
    </div>
    <div
      className="col-6"
    >
      <div
        id="GrossIncome_ID"
      >
        $1,500.00
      </div>
    </div>
  </div>
  <div
    className="row grey-row"
  >
    <div
      className="col-6"
    >
      <div
        id="IncomeTax_Label"
      >
        <strong>
          Income Tax
        </strong>
      </div>
    </div>
    <div
      className="col-6"
    >
      <div
        id="IncomeTax_ID"
      >
        $4,500.00
      </div>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col-6"
    >
      <div
        id="NetIncome_Label"
      >
        <strong>
          Net Income
        </strong>
      </div>
    </div>
    <div
      className="col-6"
    >
      <div
        id="NetIncome_ID"
      >
        $3,000.00
      </div>
    </div>
  </div>
  <div
    className="row grey-row"
  >
    <div
      className="col-6"
    >
      <div
        id="Super_Label"
      >
        <strong>
          Super
        </strong>
      </div>
    </div>
    <div
      className="col-6"
    >
      <div
        id="Super_ID"
      >
        $140.00
      </div>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col-6"
    >
      <div
        id="Pay_Label"
      >
        <strong>
          Pay
        </strong>
      </div>
    </div>
    <div
      className="col-6"
    >
      <div
        id="Pay_ID"
      >
        $750.00
      </div>
    </div>
  </div>
  <div
    className="row justify-content-end"
  >
    <div
      className="col-6"
    >
      <button
        className="btn btn-primary pay-btn"
        id="payButton"
        onClick={[Function]}
      >
        Pay
      </button>
    </div>
  </div>
</div>
`;
